[flow] Export type Props for composability

rosskevin opened the pull request:
Change all `^type Props` to `export type Props` to allow for external composability using material-ui's defined types. e.g. If I want to enhance a `TextField` in some way for my own library, it is easiest to import the existing `Props` and add my own to create the enhanced Props for my custom component.

## Important pattern of use change
Introduce `AllProps` type and remove intersection from externalized Props type.

External use of exported `type Props = DefaultProps & {...}` was hindered as it was an intersection that mixed in the `DefaultProps`.  This prevented external use of composite types when using `$Exact<Props>` a.k.a. `{|Props|}`.  This commit keeps everyting in isolation and exported Props are now purely as intended/used in react.

---
- [x] PR has tests / docs demo, and is linted.
- [x] Commit and PR titles begin with [ComponentName], and are in imperative form: "[Component] Fix leaky abstraction".
- [x] Description explains the issue / use-case resolved, and auto-closes the related issue(s) (http://tr.im/vFqem).

rosskevin committed:
[flow] export type Props

rosskevin commented:
Transpiled output for each file now adds the export, here is an example of the additional code from `TextField`:

```javascript
if (typeof exports !== 'undefined') Object.defineProperty(exports, 'babelPluginFlowReactPropTypes_proptype_Props', {
  value: {
    autoComplete: require('prop-types').string,
    autoFocus: require('prop-types').bool,
    className: require('prop-types').string,
    defaultValue: require('prop-types').string,
    disabled: require('prop-types').bool,
    error: require('prop-types').bool,
    FormHelperTextProps: require('prop-types').object,
    fullWidth: require('prop-types').bool,
    helperText: require('prop-types').oneOfType([require('prop-types').string, typeof babelPluginFlowReactPropTypes_proptype_Element === 'function' ? babelPluginFlowReactPropTypes_proptype_Element : require('prop-types').shape(babelPluginFlowReactPropTypes_proptype_Element)]),
    helperTextClassName: require('prop-types').string,
    id: require('prop-types').string,
    inputClassName: require('prop-types').string,
    InputClassName: require('prop-types').string,
    InputLabelProps: require('prop-types').object,
    inputProps: require('prop-types').object,
    InputProps: require('prop-types').object,
    inputRef: require('prop-types').func,
    label: require('prop-types').oneOfType([require('prop-types').string, typeof babelPluginFlowReactPropTypes_proptype_Element === 'function' ? babelPluginFlowReactPropTypes_proptype_Element : require('prop-types').shape(babelPluginFlowReactPropTypes_proptype_Element)]),
    labelClassName: require('prop-types').string,
    multiline: require('prop-types').bool,
    name: require('prop-types').string,
    placeholder: require('prop-types').string,
    required: require('prop-types').bool,
    rootRef: require('prop-types').func,
    rows: require('prop-types').oneOfType([require('prop-types').string, require('prop-types').number]),
    rowsMax: require('prop-types').oneOfType([require('prop-types').string, require('prop-types').number]),
    type: require('prop-types').string,
    value: require('prop-types').oneOfType([require('prop-types').string, require('prop-types').number]),
    margin: require('prop-types').oneOf(['none', 'dense', 'normal'])
  },
  configurable: true
});
```

My goal is to merely make this available from a flow perspective, not necessarily to export the generated props and make them available in any kind of runtime.  I don't mind the package growth; I'm just not sure how to stop it from happening. 

/cc @oliviertassinari

oliviertassinari commented:
Your change in increasing the payload of the lib by 3kgb gzip, that's a 3% increase. Is the feature worth it?

rosskevin commented:
I should also note that this code is almost an exact duplicate of the transpiled code below too:

```javascript
TextField.propTypes = process.env.NODE_ENV !== "production" ? {
  autoComplete: require('prop-types').string,
  autoFocus: require('prop-types').bool,
  className: require('prop-types').string,
  defaultValue: require('prop-types').string,
  disabled: require('prop-types').bool,
  error: require('prop-types').bool,
  FormHelperTextProps: require('prop-types').object,
  fullWidth: require('prop-types').bool,
  helperText: require('prop-types').oneOfType([require('prop-types').string, typeof babelPluginFlowReactPropTypes_proptype_Element === 'function' ? babelPluginFlowReactPropTypes_proptype_Element : require('prop-types').shape(babelPluginFlowReactPropTypes_proptype_Element)]),
  helperTextClassName: require('prop-types').string,
  id: require('prop-types').string,
  inputClassName: require('prop-types').string,
  InputClassName: require('prop-types').string,
  InputLabelProps: require('prop-types').object,
  inputProps: require('prop-types').object,
  InputProps: require('prop-types').object,
  inputRef: require('prop-types').func,
  label: require('prop-types').oneOfType([require('prop-types').string, typeof babelPluginFlowReactPropTypes_proptype_Element === 'function' ? babelPluginFlowReactPropTypes_proptype_Element : require('prop-types').shape(babelPluginFlowReactPropTypes_proptype_Element)]),
  labelClassName: require('prop-types').string,
  multiline: require('prop-types').bool,
  name: require('prop-types').string,
  placeholder: require('prop-types').string,
  required: require('prop-types').bool,
  rootRef: require('prop-types').func,
  rows: require('prop-types').oneOfType([require('prop-types').string, require('prop-types').number]),
  rowsMax: require('prop-types').oneOfType([require('prop-types').string, require('prop-types').number]),
  type: require('prop-types').string,
  value: require('prop-types').oneOfType([require('prop-types').string, require('prop-types').number]),
  margin: require('prop-types').oneOf(['none', 'dense', 'normal'])
} : {};
```

So BOTH of these are in the transpiled file.

oliviertassinari commented:
But I don't get it, why do we even have this extra code? Shouldn't flow be removed from the published package?

rosskevin commented:
I think exporting the types is important for external composability, but the large amount of duplication is hard to swallow in the transpiled file.  This is really only necessary in the `*.js.flow` file.

Flow is fully removed, but the `babel-plugin-flow-react-prop-types` plugin is generating this output.

oliviertassinari commented:
I don't think that code duplicated is going to help us here. The prop-types are removed in production.

rosskevin commented:
We would have to wrap both, and alter your production removal plugin.

oliviertassinari commented:
Or should we fix `babel-plugin-flow-react-prop-types` instead?

rosskevin commented:
The right fix is the plugin needs to declare 

```javascript
var propTypes = { /* ... */ }; 
TextField.propTypes = propTypes;
if (typeof exports !== 'undefined') Object.defineProperty(exports, 'babelPluginFlowReactPropTypes_proptype_Props', propTypes);
```

oliviertassinari commented:
From what I can see, the size-limit reported take into account the removal of the prop-types with [this line](https://github.com/ai/size-limit/blob/ac674e160319369c8f7a9e462c487e2cef0c4cf9/index.js#L38).

rosskevin commented:
I'm looking at the plugin now.  Not sure it will be quite so easy to get to what I think is the right solution.  I'll try to figure it out, but it would certainly be easier to just wrap/omit the export too.

rosskevin committed:
update flow prop type generation plugin

rosskevin commented:
@oliviertassinari - I've PR'd the fix to the prop type generation plugin and updated it here.

The remove-prop-types wrapping might need to be changed.  The current use is correct, but also when prop types are exported, the code generated now uses a variable:

```javascript
var babelPluginFlowReactPropTypes_proptype_Props = { /* ... */ }
if (typeof exports !== 'undefined') Object.defineProperty(exports, 'babelPluginFlowReactPropTypes_proptype_Props', {
  value: babelPluginFlowReactPropTypes_proptype_Props,
  configurable: true
});
TextField.propTypes = process.env.NODE_ENV !== "production" ? babelPluginFlowReactPropTypes_proptype_Props : {};
```

Will you take a look at it?

oliviertassinari commented:
I'm happy to see https://github.com/brigand/babel-plugin-flow-react-proptypes/pull/121 :).
Regarding, `remove-prop-types` I don't see the link here. The plugin responsability is to remove the reference between a component and his propTypes. Any further optimization is left to dead code elimination tools, like uglifyjs. If the propTypes are exported, uglifyjs wont' remove anything.
1. I don't think that the flow plugin should export the prop-types
2. We can always write a custom plugin for that use case.

rosskevin commented:
1. I don't think that the flow generation plugin should be opinionated enough to _not_ export props.  This _could_ be used for reusing shapes etc (something we aren't doing)
2. I agree on dead code elimination, we just need to kill the export
3. I think a custom kill of the export would work

Are you interested in working on a custom transform to kill the export?  My babel transform skills are meager at best.

rosskevin commented:
@oliviertassinari I just had a thought - I might be able to slip in the omit export into the generation plugin as an option.  I'll see if I can get that done first.

rosskevin committed:
[flow] `classes` is optional as an external prop, but always present inside a withStyles enhanced component

rosskevin committed:
[flow] rebased, remainder of `classes` treatment as optional external to components

rosskevin commented:
I'm actively using  the exported types and trying to resolve errors presented by external use.  I'll remove the WIP tag when ready.

rosskevin committed:
[flow] introduce `AllProps` type and remove intersection from externalized Props type

External use of exported `type Props` was hindered as it was an intersection that mixed in the `DefaultProps`.  This prevented external use of composite types when using $Exact<Props> or {|Props|}.  This commit keeps everyting in isolation and exported Props are now pure.

rosskevin committed:
babel rc option for optimization

rosskevin committed:
[Typography] fix prop vs default prop headlineMapping

rosskevin committed:
[babelrc] temporarily remove options breaking build

rosskevin committed:
upgrade babel-plugin-transform-react-remove-prop-types and flow

rosskevin committed:
[ButtonBase] className is optional

oliviertassinari commented:
I have tried, it's working. The propTypes are no longer referenced. I wish they weren't generated at all, but that's another story. We should be able to move forward.

```diff
     "babel-plugin-transform-react-constant-elements": "^6.23.0",
-    "babel-plugin-transform-react-remove-prop-types": "^0.4.7",
+    "babel-plugin-transform-react-remove-prop-types": "^0.4.8",
     "babel-plugin-transform-runtime": "^6.23.0",
```

oliviertassinari commented:
One last thing, please "Squash and merge" over "Create a merge commit", it's a mess in the history for create the Change log otherwise.

rosskevin committed:
update the transform remove prop types version
